### DISCUSSION TITLE
Allow empty multi-select-list and multi-select-hash

### DISCUSSION THREAD

**Discussion by innovate-invent** [jmespath-community/jmespath.spec#93] [Ideas]
There are a few situations where it would be useful to instantiate an empty array/object.

This is particularly important for reduce() to be able to pass an empty value as the init expression.
This could also be used with the arithmetic operators for a fallback operand, ie: `(a || {}) + {foo: 5}` in the event that `a` is undefined or otherwise falsey. A more advanced example: `(type(a) == 'object' && a || {}) + {foo: 5}`.

I believe there is room in the grammar to allow `{}` to represent an empty object, but `[]` currently refers to the flatten operator.

We could have to_array() accept zero arguments and return an empty array.
``[`null`][]`` is an existing and valid way to get an empty array. This could be shortened to `[a][]` assuming 'a' is undefined in the context.
Are there any better ways to do this?

The proposed grammar change is:
```abnf
multi-select-hash = "{" [ keyval-expr *( "," keyval-expr ) ] "}" 
multi-select-list = "[" ( expression *( "," expression ) / " " ) "]" 
```

**Comment by springcomp** [comment]
- You can instantiate an empty array with an empty `multi-select-list`, like so: `[  ]` (note at least a space between the opening and closing brackets, to prevent this from being tokenized to a `flatten-expression`.
   For instance, `` search( length([  ]), [1, 2, 3] ) -> 0 ``

- We could indeed extend `` {} `` to be an empty `hash-wildcard-expression` object. Although that is currently possible with `raw-string` literals.
   For instance, `` search( `{}`, [1, 2, 3] ) -> {} ``

**Reply by innovate-invent** [reply]
Ha!, I don't know why raw string literals completely escaped me. I would prefer to move in a direction that allows us to eliminate them from the grammar though.

The space in the `[ ]` is a cool hack, I like it. It does violate the grammar rules:
```abnf
multi-select-list = "[" ( expression *( "," expression ) ) "]" 
```
The first expression is required. I will add that to the proposal.

**Reply by springcomp** [reply]
You are right. I wondered if that was correct when I stumbled into it.
We definitely need to standardize this as part of the JEP.

Would you feel like writing the JEP draft here ?

I think it would:
- allow empty `multi-select-list`.
- specify a standard symmetric empty `multi-select-hash` object. like `{}`.

**Reply by springcomp** [reply]
I think empty `multi-select-hash` expression can be [standardized in this JEP](https://github.com/jmespath-community/jmespath.spec/discussions/19#discussion-3913995).

**Comment by springcomp** [comment]
This has been included in [Compact Syntax for multi-select-hash](https://github.com/jmespath-community/jmespath.spec/discussions/19#discussion-3913995)
